algebra-core/src/curves/bucketed_add.rs

```
@@ -0,0 +1,256 @@
+                                .push((assign_vec[2 * j] as usize, assign_vec[2 * j + 1] as usize));
+                        }
+                        if len % 2 == 1 {
+                            index[idx + new_len] = assign_vec[len - 1];
+                        }
+                        assign_hash.remove(&bucket);
+                    } else {
+                        for j in 0..len / 2 {
+                            assign_vec[j] = assign_vec[2 * j];
+                            instr_row
+                                .push((assign_vec[2 * j] as usize, assign_vec[2 * j + 1] as usize));
+                        }
+                        if len % 2 == 1 {
+                            assign_vec[new_len - 1] = assign_vec[len - 1];
+                        }
+                    }
+                } else {
+                    for j in 0..len / 2 {
+                        index[idx + j + 1] = index[idx + 2 * j + 1];
+                        instr_row.push((
+                            index[idx + 2 * j + 1] as usize,
+                            index[idx + 2 * j + 2] as usize,
+                        ));
+                    }
+                    if len % 2 == 1 {
+                        index[idx + new_len] = index[idx + len];
+                    }
+                }
+                // New length is the ceil of (old_length / 2)
+                index[idx] = new_len as u32;
+            }
+        }
+        if instr_row.len() > 0 {
+            instr.push(instr_row);
+        }
+    }
+
+    for instr_row in instr.iter() {
+        for instr_chunk in
+            C::get_chunked_instr::<(usize, usize)>(&instr_row[..], BATCH_ADD_SIZE).iter()
+        {
+            elems[..].batch_add_in_place_same_slice(&instr_chunk[..]);
+        }
+    }
+
+    let zero = C::zero();
+    let mut res = vec![zero; buckets];
+
+    for bucket in 0..buckets {
+        if index[offset * bucket] > 1 {
+            panic!("Did not successfully reduce to_add");
```

Kinda scary, can this even happen?

---

Technically, if the code is correct, no. It was used for debugging. And if the code is wrong in some subtle way, at least if could catch a silent error. I'll change to a debug assert so the compiler optimizes it away